Dynamic filtering support

Auwate opened the pull request:
# Description

Tackles #12 to support the network operators team while providing reliable support for the `Kytos` team.

More information can be found in the following comments:

https://github.com/kytos-ng/kytos/pull/557#issuecomment-2830777700
https://github.com/kytos-ng/kytos/pull/557#issuecomment-2839444097
https://github.com/kytos-ng/kytos/pull/557#issuecomment-2842629449

# Changes

- Updated README.md
- Added /models/filter.py
- Added endpoints to main.py
- Added /managers/regex.py

Auwate committed:
Commit:

# Changes
- README.md
- Filter.py
- Endpoints to main.py
- Regex.py

Alopalao commented on managers/regex.py:
```diff
@@ -1,49 +1,72 @@
+            if not self.rules[rule].is_mutable():
+                raise ValueError("Pattern cannot be removed.")
+            return self.rules.pop(rule)
+        except (KeyError, ValueError) as exc:
```
A lock here is needed. `self.rules.pop(rule)` should not be raising `KeyError` if you checked that it is already in the dictionary. After adding the lock you can remove the redundant re-raisings too.
Edit:
Wait confirming this.

Alopalao replied:
Hey @viniarck, `self.rules` is currently only used in the context of `async` and by how the the requests are structure there should not be a problem with race conditions. But `self.rules` could still be used in a `sync` context by commands on console or future code modifications. Which one is more recommended in this situation?:
- Acquire an `async lock` here and transform the method into async
- Use `sync` methods for the requests and use a `threading.lock()`.
- Leave it as it is?

viniarck replied:
@Alopalao, that's a good observation and good to think about it. In case case, leave as it is. 

- If it ever need to be called from the console on demand on exceptional occasions it'll still work
- `async lock` is primarily used when coordinating with one or more `await`(s)
- `threading.Lock` isn't compatible with async context as you know, so we should also stay away from it in this case

Bottom line: once it's using `async`, we should stick to stay in `async` land as much as possible in terms of endpoints and events. For things that might be used in the console, that's fine, since console is just exceptional usage on demand in terms of low-ish level commands and won't blow up.

Alopalao commented on models/filter.py:
```diff
@@ -0,0 +1,67 @@
+        """
+        return self.mutable
+
+    def summarize(self) -> dict[str, str]:
```
The name is good but we already have a name for this kind of method. Please change it to `as_dict` since it is more aligned with other NApps implementations.

Alopalao commented on managers/regex.py:
```diff
@@ -1,49 +1,72 @@
-        regex = escaped.replace("\\*", ".*")
-        return regex
+        try:
+            if rule in self.rules:
```
`self.rules` is a shared collection you would need a lock to not run into issues with tasks.
Edit:
Lets wait for a response in this [comment](https://github.com/amlight/kafka_napp/pull/13#discussion_r2082141360).

Alopalao reviewed (commented):
It is shaping up well. Left some comments about multitasking that need to be addressed.
Also if you could add an `openapi.yml` so the content of the requests are documented. For example we have one for [mef_eline](https://github.com/kytos-ng/mef_eline/blob/master/openapi.yml).
Sorry about being late to review.

viniarck commented on main.py:
```diff
@@ -66,3 +69,68 @@ async def handle_events(self, event: KytosEvent):
+        Requires 1 provided value in the content section called "pattern". Accepts 1 optional
+        value called "description"
+        """
+        log.info("Request received at /v1/create")
```
`log.info` at request handlers only fits well if it's a endpoint that might not get called too frequently or when indeed it's logging state that's relevant for correlation. I'd recommend either only use `log.debug` if you need it or remove it. The status code and response should suffice for this case, and `uvicorn` also already logs requests to that ends up being a bit redundant in terms of logs.

Auwate replied:
Got it, I'll modify it for only debugging purposes

viniarck commented on main.py:
```diff
@@ -66,3 +69,68 @@ async def handle_events(self, event: KytosEvent):
+                }
+            )
+        except (ValueError, RegexException) as exc:
+            return JSONResponse({"status": "400", "response": exc})
```
Any non success code in our codebase when handling a request we recommend to raise `HTTPException` here's an example https://github.com/kytos-ng/mef_eline/blob/master/main.py#L214. This also simplifies error handling for clients (front-end and any other client) since the response fields will use standardized keys.

viniarck commented on main.py:
```diff
@@ -66,3 +69,68 @@ async def handle_events(self, event: KytosEvent):
             )
         )
+
+    @rest("v1/create", methods=['POST'])
```
Usually a `create` in the URL with `POST` is a bit redundant it should ideally use the resource in its URL. Also, `/v1/list` and `/v1/delete` `POST` aren't ideal either. Let's review the endpoints spec, and I recommend you also include an openapi.yml spec. Usually no action verbs are included in the endpoints. You can consider using a `"v1/filter"` or `"v1/event_filter"` or something along these lines, think about it, something related to the underlying managed resource, `filter` in this case being a noun

viniarck commented on main.py:
```diff
@@ -66,3 +69,68 @@ async def handle_events(self, event: KytosEvent):
+
+            return JSONResponse(
+                {
+                    "status": "200",
```
In success case we avoid having a redundant status code info in the response, similar to what we have in other NApps https://github.com/kytos-ng/mef_eline/blob/master/main.py#L217. Also, "response" key is a bit redundant in the action response.

viniarck reviewed (changes requested):
Anticipating some of the review since my feedback was asked. Glad to see how it's progressing. The final review I'll do in the blueprint and on other PRs that eventually will be sent to kytos-ng org.

Auwate committed:
Commit:

# Changes
- Updated main.py to reflect best practices in REST API usage
- Created openapi.yml doc
- Updated regex.py to be completely async-oriented
- Updated filter.py to use as_dict()
- Updated unit tests

Auwate commented:
# Summary

- Updated `main.py` endpoints to follow REST API best practices
  - GET endpoint is now `/v1/filters`
  - POST endpoint is now `/v1/filters`
  - DELETE endpoint it now `/v1/filters/{filter_id`
- Updated `main.py` endpoints to return standardized `JSON` data
  - GET returns a list of filter objects
  - POST returns the newly created filter object
  - DELETE returns the deleted filter object
- Updated setup.py to use the new name `kafka_events`
  - Was changed as sending a request would have been to `/api/kytos/kafka_napp` instead.
- Updated `filter.py` to use `as_dict()` to maintain consistency
- Updated `regex.py` to be completely `async-oriented` to maintain the shift towards async and not block the rest of Kytos
- Updated unit testing to reflect changes in behavior and implementation
- Created `openapi.yml` document to reflect all endpoint behaviors, what they expect, and what they return.

Auwate committed:
Commit:

# Changes
- Updated README.md to reflect new changes in REST endpoint usage

Alopalao commented on managers/regex.py:
```diff
@@ -1,59 +1,157 @@
+        except (KeyError, ValueError) as exc:
+            raise exc
+
+    async def as_dict(self, filter_id: str, filter_obj: Filter) -> dict[str, str]:
```
`as_dict`  historically does not have any required arguments. I fell like it should be like that if I ever wanted only field I would write it like `my_object.as_dict()["desired_field"]`.
This method purpose seems too redundant, the ID does not need to be in its own field it could be just the key and be understood.

Auwate replied:
Got it! I'll keep `as_dict` as a representation of an object and call that representation when possible from the service layer.

Alopalao commented on managers/regex.py:
```diff
@@ -1,59 +1,157 @@
+        Raises ValueError if the given pattern immutable
+        """
+        try:
+            return await self.as_dict(filter_id, await self._delete_filter(filter_id))
```
With the removal of `RegexOperations.as_dict()` we could also get rid of this method. No need to create another method to just call another one.

Alopalao commented on main.py:
```diff
@@ -55,14 +59,89 @@ async def handle_events(self, event: KytosEvent):
+            status: int = 200
+
+            return JSONResponse(
+                    filter_obj,
```
No need to specify the ID in a new field, it could just be the key. E.g.:
```
            deleted_filter: dict[str, str] = await self._rule_ops.delete_filter(filter_id)
            status: int = 200

            return JSONResponse(
                    {filter_id: deleted_filter},
```

Alopalao commented on managers/regex.py:
```diff
@@ -1,59 +1,157 @@
+        return [
+            await self.as_dict(
+                filter_id, filter_obj
+            ) for filter_id, filter_obj in self.rules.items()
```
Just create the list directly from `self.rules`.

Alopalao commented on main.py:
```diff
@@ -55,14 +59,89 @@ async def handle_events(self, event: KytosEvent):
+            status: int = 201
+
+            return JSONResponse(
+                filter_obj,
```
Same as in `delete_filter`, the ID could be the key no need to specify it.

Alopalao commented on setup.py:
```diff
@@ -24,7 +24,7 @@
 NAPP_USERNAME = 'kytos'
 NAPP_VERSION = '1.0.0'
-NAPP_NAME = "kafka_napp"
+NAPP_NAME = "kafka_events"
```
Let's keep consistency with the names. If this is the final name, then it will also need to be change in `kytos.json`. Do not forget to change the name of the repo too.

Auwate replied:
Yes, I tried keeping it consistent but a few must have slipped under the radar.

Alopalao reviewed (commented):
Good job, I left a few comments where code could be reduced. Please do not forget to change the name to of this repo if the final name ends up as `kafka_events`. Also change the name of the napp in `kytos.json`.

Auwate committed:
Commit:

# Removed 'as_dict' from regex.py in favor of entity-level as_dict usage
# Updated APIs to return the UUID as the key to the object
# Updated OpenAPI documentation to support new changes

Auwate commented:
> Please do not forget to change the name to of this repo if the final name ends up as kafka_events.

Once the PR is accepted, I'll ask Vinicius to change the repo name to `kafka_events`, aligning with the new name defined in `EP039`.

In addition, the new VM has been created so benchmarks are on their way soon.

Auwate committed:
Commit:

# Description:
- Updated openapi.yml documentation to reflect new changes in API responses.
  - The object key should be a dynamic UUID, removing the 'id' field from the inner properties

Alopalao reviewed (commented):
LGTM, You can go ahead and add the missing unit tests and fix the ones that are failing. I think Vinicius wanted to give a final review, just give him some heads up.

Auwate committed:
Commit:

# Description:
- Updated unit testing to reflect new changes
  - The tests were failing due to changes in logic within main.py and regex.py
- Added new unit test to cover new behavior in handle_events

Auwate commented:
> LGTM, You can go ahead and add the missing unit tests and fix the ones that are failing. I think Vinicius wanted to give a final review, just give him some heads up.

Thanks for the heads up on testing! I aligned each test to reflect the new changes in each and added a unit test to validate `handle_events` new behavior of ignoring tests that are found in any patterns within the RegexOperations `_rules` dictionary. It shows 19 tests are passing now.

I updated Vinicius on the progress and he will get back to us soon.

Auwate commented:
After this PR is accepted I'll ask for the repository name to be changed to `kafka_events`, as well as changing any references in testing to `kafka_events`. Currently they point to `kafka_napp` only because that's the repository name and would cause tests to fail.

viniarck reviewed (approved):
@Auwate, I did an overall review, it looks good to me. It's great that you're managing to progress but. Go ahead and land it, and I'll rename this repo as `kafka_events` as you asked.

Other than that, here's my recommendation to you:

1) Focus on finishing https://github.com/kytos-ng/kytos/pull/557 and getting it accepted. Let me know when you need a final review there.
  - It's crucial to finish there just so there's no other major questions and back and forth in code reviews in terms of requirements. For instance, when it comes to requirements, currently, it's not clear to me how filters persistency will work. Also, I look forward to the benchmark results, hopefully it'll suffice, and that's also pivotal to know that this NApp will be viable to use in prod. It's OK to keep developing other parts as you're sorting out the other ones, but just to align code review expectations without certain requirements it causes more attrition when it comes to accepting the code upstream.
2) Once the blueprint is accepted, I'll create a repo for you in the kytos-ng org, and then I'll also hook the scrutinizer CI to run unit tests on every PR, and from that point forward, send PRs there and core reviewers will review your contributions.
  - Also, start thinking about how the e2e tests will be covered too (we do ask for e2e tests in code reviews, especially for major changes, and without them we also accept local exploratory tests, but still it's yet another point to align expectations)
